### PR TITLE
handler: Expose nmstatectl stats as k8s metrics

### PR DESCRIPTION
**Pull request by qinqon**
**Is this a BUG FIX or a FEATURE ?**:
/kind enhancement

**What this PR does / why we need it**:
Now that nmstatectl is able to calculate some useful stats from network configuration [1], we can bubble them up and expose them as k8s metrics so k-nmstate users can digg on them using prometheus, graphana or the like.

This change add a new "Features" under nnce Status with the output of
`nmstatectl st` and also create a new deployment `nmstate-metrics` that
will gather the NNCEs features and reflecta that at a cluster wide
gaugue prometheus metric.

This is an example of nmstate feature stat

```
kubernetes_nmstate_features_applied{name="dhcpv4-custom-hostname"} 1
```

Depends on nmstate 2.2.20, looks like it's build but still not present at centos 9 stream
- https://kojihub.stream.rdu2.redhat.com/koji/buildinfo?buildID=41534

[1] https://github.com/nmstate/nmstate/pull/2420

TODO:
- [ ] Compare old and new nncp stats to be able to decrease conunters.

**Release note**:

```release-note
Expoxe statistics generated from `nmstatectl stats`
```

### CODE REVIEW

**Comment by sradco** [comment]
@machadovilaca please see if you can help in reviewing this PR.

File: pkg/monitoring/metrics.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	ApplyTopologyTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "nmstate_apply_topology_total",
+		},
+		[]string{"name"},
+	)
+
+	ApplyFeaturesTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "nmstate_apply_features_total",
```
Comment (sradco) [inline pkg/monitoring/metrics.go:34]: @qinqon You are missing a help text for the metrics.

File: pkg/monitoring/metrics.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+package monitoring
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	ApplyTopologyTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "nmstate_apply_topology_total",
```
Comment (sradco) [inline pkg/monitoring/metrics.go:27]: If I understand currently, This metric counts the number of times the topology was applied.
If that is correct I would name it `nmstate_topology_applied_total`.

File: pkg/monitoring/metrics.go (hunk 1 above)
Comment (sradco) [reply pkg/monitoring/metrics.go:34]: If I understand currently, This metric counts the number of times the features were applied.
If that is correct I would name it `nmstate_features_applied_total`.

File: pkg/monitoring/metrics.go (hunk 2 above)
Comment (sradco) [reply pkg/monitoring/metrics.go:27]: Please add a help test. Its visible in the Grafana and OCP UI when searching for the metric. Its also useful for generation automated metrics documentation, like we have in kubevirt.

Comment (sradco) [reply pkg/monitoring/metrics.go:27]: We also have a metrics name linter that I would suggest you use in order to validate the metrics names meet the best practices. We added it to the PR tests, so issues will be detected early.

**Comment by avlitman** [comment]
> @machadovilaca @avlitman please see if you can help in reviewing this PR.

+1 Will take a look asap (probably tomorrow).

File: pkg/monitoring/metrics.go (hunk 2 above)
Comment (qinqon) [reply pkg/monitoring/metrics.go:27]: @sradco do you have a pointer to that linter to integrate it ?

File: test/e2e/handler/metrics.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"strings"
+
+	"github.com/nmstate/kubernetes-nmstate/test/cmd"
+	"github.com/nmstate/kubernetes-nmstate/test/runner"
+)
+
+func getMetrics(token string) string {
+	bearer := "Authorization: Bearer " + token
+	return runner.RunAtFirstHandlerPod("curl", "-s", "-k", "--header", bearer, "https://127.0.0.1:8443/metrics")
```
Comment (machadovilaca) [inline test/e2e/handler/metrics.go:12]: you can use `metrics.DefaultBindAddress` here too instead of the port hardcoded value

File: test/e2e/handler/metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+			})
+			AfterEach(func() {
+				updateDesiredStateAndWait(linuxBrAbsent(bridge1))
+				resetDesiredStateForNodes()
+			})
+			It("should report a metrics with proper counter increased", func() {
+				token, err := getPrometheusToken()
+				Expect(err).ToNot(HaveOccurred())
+				metrics := getMetrics(token)
+				Expect(findMetric(metrics, "fake metrics")).ToNot(BeEmpty())
```
Comment (machadovilaca) [inline test/e2e/handler/metrics_test.go:22]: is there a metric named "fake metrics", where does it come from?

File: test/e2e/handler/metrics.go (hunk 3 above)
Comment (machadovilaca) [inline test/e2e/handler/metrics.go:12]: some error handling could make sense here, wdyt?

File: pkg/nmstatectl/nmstatectl.go [hunk 5]
```diff
@@ -94,3 +95,24 @@ func Rollback() error {
...
+type Stats struct {
+	Topology []string `json:"topology"`
+	Features []string `json:"features"`
+}
+
+func Statistic(desiredState nmstate.State) (*Stats, error) {
+	statsOutput, err := nmstatectlWithInput(
+		[]string{"st", "-"},
+		string(desiredState.Raw),
+	)
```
Comment (machadovilaca) [inline pkg/nmstatectl/nmstatectl.go:108]: can you help me find some documentation or example for the output of this?

File: deploy/handler/operator.yaml [hunk 6]
```diff
@@ -279,6 +281,28 @@ spec:
...
+          ports:
+          - containerPort: 8443
+            name: metrics
+            protocol: TCP
+          resources:
+            requests:
+              cpu: 10m
+              memory: 20Mi
+          terminationMessagePolicy: FallbackToLogsOnError
+
```
Comment (machadovilaca) [inline deploy/handler/operator.yaml:305]: why is this needed in this PR?

File: deploy/handler/role.yaml [hunk 7]
```diff
@@ -118,6 +128,18 @@ rules:
...
+  resources:
+  - tokenreviews
+  verbs:
+  - create
+- apiGroups:
+  - authorization.k8s.io
+  resources:
+  - subjectaccessreviews
+  verbs:
+  - create
```
Comment (machadovilaca) [inline deploy/handler/role.yaml:142]: why is this needed in this PR?

File: test/e2e/handler/metrics_test.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	Context("when desiredState is configured", func() {
+		Context("with a state that generate metrics", func() {
+			BeforeEach(func() {
+				updateDesiredStateAndWait(linuxBrUpNoPorts(bridge1))
+			})
+			AfterEach(func() {
+				updateDesiredStateAndWait(linuxBrAbsent(bridge1))
+				resetDesiredStateForNodes()
+			})
+			It("should report a metrics with proper counter increased", func() {
```
Comment (machadovilaca) [inline test/e2e/handler/metrics_test.go:18]: the test does not reflect what this description proposes

File: pkg/monitoring/metrics.go (hunk 2 above)
Comment (avlitman) [reply pkg/monitoring/metrics.go:27]: The linter we use for kubevirt is here: https://github.com/kubevirt/monitoring/tree/main/test/metrics/prom-metrics-linter, see the readme as well it's basically the [promlint](https://github.com/prometheus/client_golang/blob/v1.16.0/prometheus/testutil/promlint/promlint.go), and we added that the metric should start with kubevirt_hco/cdi/ssp/hpp/..)

Counter do need to end with total, maybe the start should be `kubernetes_nmstate`?

File: test/e2e/handler/metrics.go (hunk 3 above)
Comment (qinqon) [reply test/e2e/handler/metrics.go:12]: I am not sure the address here is not `:8080`

Comment (machadovilaca) [reply test/e2e/handler/metrics.go:12]: you also have `8443` configured in the service you created

File: test/e2e/handler/metrics_test.go (hunk 4 above)
Comment (qinqon) [reply test/e2e/handler/metrics_test.go:22]: This is wrong, will fix it.

File: test/e2e/handler/metrics.go (hunk 3 above)
Comment (qinqon) [reply test/e2e/handler/metrics.go:12]: The error is asserted inside with gomega.

File: deploy/handler/operator.yaml (hunk 6 above)
Comment (qinqon) [reply deploy/handler/operator.yaml:305]: You need a kube-rbac-proxy to allow access to the metrics from prometheus

File: test/e2e/handler/metrics_test.go (hunk 8 above)
Comment (qinqon) [reply test/e2e/handler/metrics_test.go:18]: Have to fix it.

File: deploy/handler/role.yaml (hunk 7 above)
Comment (qinqon) [reply deploy/handler/role.yaml:142]: Need by prometheus integration at nmstate-handler

File: pkg/monitoring/metrics.go (hunk 2 above)
Comment (avlitman) [reply pkg/monitoring/metrics.go]: I agree with 'nmstate_topology_applied_total'

**Comment by qinqon** [comment]
@avlitman @machadovilaca @sradco can you take another look ?

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 9]
```diff
@@ -223,6 +224,18 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(_ context.Context,
...
+	stats, err := nmstatectl.Statistic(enactmentInstance.Status.DesiredState)
+	if err != nil {
+		log.Error(err, "failed retrieving stats for desired state")
+	} else {
+		for _, t := range stats.Topology {
+			monitoring.ApplyTopologyTotal.WithLabelValues(t).Inc()
+		}
+		for _, f := range stats.Features {
+			monitoring.ApplyFeaturesTotal.WithLabelValues(f).Inc()
+		}
```
Comment (machadovilaca) [inline controllers/handler/nodenetworkconfigurationpolicy_controller.go:236]: Per https://github.com/nmstate/kubernetes-nmstate/pull/1221/files#r1431131839, as I understand, I'm afraid this will have unbound cardinality as it will contain network state information like hostnames/interfaces/address/ports, is that correct?

File: pkg/monitoring/metrics.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	ApplyTopologyTotal = prometheus.NewCounterVec(
+		ApplyTopologyTotalOpts,
+		[]string{"name"},
+	)
+
+	ApplyFeaturesTotal = prometheus.NewCounterVec(
+		ApplyFeaturesTotalOpts,
+		[]string{"name"},
+	)
+	counterOpts = []prometheus.CounterOpts{
```
Comment (machadovilaca) [inline pkg/monitoring/metrics.go:46]: nit: new line

File: test/e2e/handler/metrics.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+*/
+
+package handler
+
+import (
+	"strings"
+
+	"github.com/nmstate/kubernetes-nmstate/test/cmd"
+	"github.com/nmstate/kubernetes-nmstate/test/runner"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
```
Comment (machadovilaca) [inline test/e2e/handler/metrics.go:25]:
```suggestion
	"sigs.k8s.io/controller-runtime/pkg/metrics"
	
	"github.com/nmstate/kubernetes-nmstate/test/cmd"
	"github.com/nmstate/kubernetes-nmstate/test/runner"
```

File: test/e2e/handler/metrics_test.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    state: up
+    ipv4:
+      enabled: true
+      dhcp: true
+      dhcp-custom-hostname: foo
+    bridge:
+      options:
+        stp:
+          enabled: false
+      port: []
```
Comment (machadovilaca) [inline test/e2e/handler/metrics_test.go:45]: move this to a var in the surrounding context to make the test easier to read

File: test/e2e/handler/metrics_test.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+`, bridge1)))
+			})
+			AfterEach(func() {
+				updateDesiredStateAndWait(linuxBrAbsent(bridge1))
+				resetDesiredStateForNodes()
+			})
+			It("should report a metrics with proper counter increased", func() {
```
Comment (machadovilaca) [inline test/e2e/handler/metrics_test.go:52]: nit: new line

File: test/e2e/handler/metrics_test.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+      dhcp: true
+      dhcp-custom-hostname: foo
+    bridge:
+      options:
+        stp:
+          enabled: false
+      port: []
+`, bridge1)))
+			})
+			AfterEach(func() {
```
Comment (machadovilaca) [inline test/e2e/handler/metrics_test.go:48]: nit: new line

**Review by machadovilaca — CHANGES_REQUESTED** [review]

**Comment by sradco** [comment]
@qinqon I propose you to also add docs generator for the operator. I think this is really useful.

We have the automation, so that when the user adds a PR with a new metric, the test runs and checks if the metric is already documented. If not the user is asked to run make generate and this automatically updated the PR with the change to the metrics.md file with the new metric, description and type.

See an example to the metrics.md file here
https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/metrics.md
and the docs generator is here
https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/tools/metricsdocs/metricsdocs.go
(Note: we plan to move it to /monitoring/tools/ )

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go (hunk 9 above)
Comment (qinqon) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:236]: > Per https://github.com/nmstate/kubernetes-nmstate/pull/1221/files#r1431131839, as I understand, I'm afraid this will have unbound cardinality as it will contain network state information like hostnames/interfaces/address/ports, is that correct?

The cardinality is not unbound, is not like we are going to count "mycluster1" and "eth1" the topology and is a conbination of enumerations (that's a little unbound but not that much) and the features is just an a list of enums

This can be a problem ?

Comment (sradco) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:236]: Yes, labels must have a limited values set. Except for specific labels like VM names, pod names, namespaces and other resources etc.
See note: https://prometheus.io/docs/practices/naming/#labels

**Comment by qinqon** [comment]
> @qinqon I propose you to also add docs generator for the operator. I think this is really useful.
> 
> We have the automation, so that when the user adds a PR with a new metric, the test runs and checks if the metric is already documented. If not the user is asked to run make generate and this automatically updated the PR with the change to the metrics.md file with the new metric, description and type.
> 
> See an example to the metrics.md file here https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/metrics.md and the docs generator is here https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/tools/metricsdocs/metricsdocs.go (Note: we plan to move it to /monitoring/tools/ )

@sradco introducing it, add a lot of golang dependencies to the project, I am not sure about, maybe we can do this at follow up.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go (hunk 9 above)
Comment (qinqon) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:236]: so this is an example of the nmstate stdats 
```
topology:
- static_ip4,static_ip6 -> linux-bridge -> bond -> ethernet
- static_ip4,static_ip6 -> vlan -> linux-bridge -> bond -> ethernet
features:
- sriov
- mac-based-identifier
```

For "features" is clear that is limited on the enum variants, maybe the main problem would be "topology".

@sradco @machadovilaca do you know what kind of limits do we have for labels bound ?

**Comment by qinqon** [comment]
@sradco @machadovilaca I have convert the Counter to Guague and decrease if topology/feature is no longer in use, can you take another look to see if everything is ok from a monitoring perspective ?

File: test/e2e/handler/metrics_test.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+				updateDesiredStateAndWait(linuxBrAbsent(bridge1))
+				resetDesiredStateForNodes()
+			})
+			It("should report a metrics with proper counter increased", func() {
+
+				token, err := getPrometheusToken()
+				Expect(err).ToNot(HaveOccurred())
+				metrics := getMetrics(token)
```
Comment (sradco) [inline test/e2e/handler/metrics_test.go:61]: @qinqon This is not a reasonable label value. Lets discus this again.
/hold

**Comment by qinqon** [comment]
We are going to steak with features for now since it has a limited bounds, we will investigate options for topology

File: hack/prom_metric_linter.sh [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+linter_image_tag="v0.0.3"
+
+# Get the metrics list
+metric_families=$(NODE_NAME=node01 go run cmd/handler/main.go --dump-metric-families)
+
+# Run the linter by using the prom-metrics-linter Docker container
+errors=$(podman run -i "quay.io/kubevirt/prom-metrics-linter:$linter_image_tag" \
+    --metric-families="$metric_families" \
+    --operator-name="kubernetes" \
```
Comment (sradco) [inline hack/prom_metric_linter.sh:31]: @qinqon  operator name should be "nmstate"
no need for sub-operator-name or it can also be "nmstate".

File: pkg/monitoring/metrics.go [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+	AppliedFeaturesOpts = prometheus.GaugeOpts{
+		Name: "kubernetes_nmstate_features_applied",
+		Help: "Number of nmstate features applied labeled by its name",
+	}
+
+	AppliedFeatures = prometheus.NewGaugeVec(
+		AppliedFeaturesOpts,
+		[]string{"name"},
+	)
+	counterOpts = []prometheus.GaugeOpts{
```
Comment (sradco) [inline pkg/monitoring/metrics.go:36]: @qinqon I think you need to update the variable name.

File: pkg/monitoring/metrics.go [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+		AppliedFeaturesOpts,
+	}
+)
+
+func Families() []pgo.MetricFamily {
+	metricFamilies := []pgo.MetricFamily{}
+	for _, counter := range counterOpts {
```
Comment (sradco) [inline pkg/monitoring/metrics.go:43]: Also here, why use "counter"?

File: test/e2e/handler/metrics_test.go [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+			BeforeEach(func() {
+
+				updateDesiredStateAndWait(linuxBridgeWithCustomHostname(bridge1))
+			})
+			AfterEach(func() {
+
+				updateDesiredStateAndWait(linuxBrAbsent(bridge1))
+				resetDesiredStateForNodes()
+			})
+			It("should report a metrics with proper counter increased", func() {
```
Comment (sradco) [inline test/e2e/handler/metrics_test.go:57]: @qinqon I believe this also need fixing since the metric is a gauge

**Comment by sradco** [comment]
@qinqon Can you please add an example to the PR description of the end metric with labels?

File: pkg/monitoring/metrics.go (hunk 18 above)
Comment (qinqon) [reply pkg/monitoring/metrics.go:43]: I forgot to change the variable name.

File: test/e2e/handler/metrics_test.go (hunk 19 above)
Comment (qinqon) [reply test/e2e/handler/metrics_test.go:57]: Yep we need an increase and decrease test with multiple features.

File: test/e2e/handler/metrics_test.go (hunk 15 above)
Comment (qinqon) [reply test/e2e/handler/metrics_test.go:61]: Is the label bounds the issue ?

**Comment by qinqon** [comment]
/retest 

```
Trying to pull registry.access.redhat.com/ubi9/ubi-minimal:latest...
Error: creating build container: copying system image from manifest list: determining manifest MIME type for docker://registry.access.redhat.com/ubi9/ubi-minimal:latest: reading manifest sha256:119ac25920c8bb50c8b5fd75dcbca369bf7d1f702b82f3d39663307890f0bf26 in registry.access.redhat.com/ubi9/ubi-minimal: received unexpected HTTP status: 502 Bad Gateway
make[1]: *** [Makefile:168: push-operator] Error 125
make[1]: Lea
```

**Comment by qinqon** [comment]
@sradco can you take another look ? I think I have cover all the comments.

**Comment by qinqon** [comment]
/hold

nmstate from "base" branch is failing a metrics tests at the "future" lane.

**Comment by qinqon** [comment]
/retest

locally with NMSTATE_PIN=future everything is fine.

**Comment by qinqon** [comment]
/hold cancel

Now is ready

File: hack/prom_metric_linter.sh [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Copyright 2023 Red Hat, Inc.
+#
+#
+set -e
+
+linter_image_tag="v0.0.3"
```
[the remark this answers is not in the document]
Comment (qinqon) [reply hack/prom_metric_linter.sh:23]: Make sense
